File: CSharp/SpaceVIL/SpaceVIL/Engine/MouseScrollProcessor.cs
using System;
using System.Collections.Generic;
using Glfw3;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal class MouseScrollProcessor
    {
        private CommonProcessor _commonProcessor;

        internal MouseScrollProcessor(CommonProcessor processor)
        {
            _commonProcessor = processor;
        }

        internal void Process(Glfw.Window glfwwnd, double dx, double dy)
        {
            if (_commonProcessor.UnderHoveredItems.Count == 0)
                return;
            Stack<Prototype> tmp = new Stack<Prototype>(_commonProcessor.UnderHoveredItems);
            while (tmp.Count > 0)
            {
                Prototype item = tmp.Pop();
                if (dy > 0 || dx < 0)
                    item.EventScrollUp?.Invoke(item, _commonProcessor.Margs);
                if (dy < 0 || dx > 0)
                    item.EventScrollDown?.Invoke(item, _commonProcessor.Margs);

                if (!item.IsPassEvents(InputEventType.MouseScroll))
                    break;
            }
            _commonProcessor.Events.SetEvent(InputEventType.MouseScroll);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/MouseScrollProcessor.java
package com.spvessel.spacevil;

import java.util.ArrayDeque;
import java.util.Deque;

import com.spvessel.spacevil.Flags.InputEventType;

final class MouseScrollProcessor {

    private CommonProcessor _commonProcessor;

    MouseScrollProcessor(CommonProcessor processor) {
        _commonProcessor = processor;
    }

    void process(long wnd, double dx, double dy) {
        if (_commonProcessor.underHoveredItems.size() == 0)
            return;

        Deque<Prototype> tmp = new ArrayDeque<>(_commonProcessor.underHoveredItems);
        
        _commonProcessor.margs.scrollValue.setValues(Math.abs(dx), Math.abs(dy));

        while (!tmp.isEmpty()) {
            Prototype item = tmp.pollLast();
            if (dy > 0 || dx < 0) {
                item.eventScrollUp.execute(item, _commonProcessor.margs);
            }
            if (dy < 0 || dx > 0) {
                item.eventScrollDown.execute(item, _commonProcessor.margs);
            }
            if (!item.isPassEvents(InputEventType.MouseScroll)) {
                break;
            }
        }
        _commonProcessor.events.setEvent(InputEventType.MouseScroll);
    }
}